### ISSUE TITLE
kvm2 tries to download via ipv6 on ipv4 only hosts

### ISSUE THREAD

**Issue by BlackMaria** [kubernetes/minikube#3654]
**BUG REPORT**
tl;dr  In the initial stages of the kvm2-vm install, the process tries to install kubernetes via ipv6 which is disabled on the host system.

As per [the discussion the other day](https://kubernetes.slack.com/archives/C1F5CT6Q1/p1549480836360600), **my** minikube kvm2 is not using ipv4, even if my OS has ipv6 disabled. Is there a way to add a hook before we pulls the packages?
* It should be noted that this may simply be my libvirt config: I had to make my own network myself, as my original issue was that minikube would complain that I dont have a default network. ( So perhaps we could add an arg to create a default network if missing? )

**Environment**:
 **minikube version:** v0.33.1

**OS:**
PRETTY_NAME="PureOS"
NAME="PureOS"
ID=pureos
HOME_URL="https://pureos.net/"
SUPPORT_URL="https://puri.sm/faq/#faq-pureosandsoftware"
BUG_REPORT_URL="https://tracker.pureos.net/"

**VM driver:**
    "DriverName": "kvm2",

**ISO version**
        "Boot2DockerURL": "file:///root/.minikube/cache/iso/minikube-v0.33.1.iso",
        "ISO": "/root/.minikube/machines/minikube/boot2docker.iso",

**kubectl version**  "v1.13.3"
**libvirt-clients** 5.0.0-1   
**qemu-kvm**  1:2.12+dfsg-3pureos1 

**What happened**:
User experience: there were a couple of unrelated bugs that came up while trying to install.
1. [ unrelated but of note ] Initially I did not have a default network in my libvirt installation, this would fail.
1. [ unrelated but of note ] I created a default network but it was supposed to be a `net-define` and not a `net-create` - perhaps we can add this to the error command

3. Finally, satisfying those, I ran `minikube  start --vm-driver=kvm2` and it failed due to ipv6 issue, see logs below.

**What you expected to happen**:

Ideally it would eventually try ipv4 or not use ipv6 at all.

**How to reproduce it** (as minimally and precisely as possible):
_I am not certain, but I assume you already have a default network on your kvm2 system, else this would have bee seen before_
* To try to reproduce what I had to do?: 
  * Delete your default network (virsh net-destroy default)  and  try `minikube  start --vm-driver=kvm2` - that should fail.
  * Create a default network, try `minikube  start --vm-driver=kvm2` and I assume it will require you to have ipv6?

**Output of `minikube logs` (if applicable)**:

``` bash
# minikube  start --vm-driver=kvm2
Starting local Kubernetes v1.13.2 cluster...
Starting VM...
Getting VM IP address...
Moving files into cluster...
Setting up certs...
Connecting to cluster...
Setting up kubeconfig...
Stopping extra container runtimes...
Starting cluster components...
E0211 19:20:02.675309    5198 start.go:376] Error starting cluster: kubeadm init: 
sudo /usr/bin/kubeadm init --config /var/lib/kubeadm.yaml --ignore-preflight-errors=DirAvailable--etc-kubernetes-manifests --ignore-preflight-errors=DirAvailable--data-minikube --ignore-preflight-errors=Port-10250 --ignore-preflight-errors=FileAvailable--etc-kubernetes-manifests-kube-scheduler.yaml --ignore-preflight-errors=FileAvailable--etc-kubernetes-manifests-kube-apiserver.yaml --ignore-preflight-errors=FileAvailable--etc-kubernetes-manifests-kube-controller-manager.yaml --ignore-preflight-errors=FileAvailable--etc-kubernetes-manifests-etcd.yaml --ignore-preflight-errors=Swap --ignore-preflight-errors=CRI 

[init] Using Kubernetes version: v1.13.2
[preflight] Running pre-flight checks
        [WARNING Swap]: running with swap on is not supported. Please disable swap
[preflight] Pulling images required for setting up a Kubernetes cluster
[preflight] This might take a minute or two, depending on the speed of your internet connection
[preflight] You can also perform this action in beforehand using 'kubeadm config images pull'
error execution phase preflight: [preflight] Some fatal errors occurred:
        [ERROR ImagePull]: failed to pull image k8s.gcr.io/kube-apiserver:v1.13.2: output: Error response from daemon: Get https://k8s.gcr.io/v2/: dial tcp: lookup k8s.gcr.io on [2001:4860:4860::8888]:53: dial udp [2001:4860:4860::8888]:53: connect: network is unreachable
, error: exit status 1
        [ERROR ImagePull]: failed to pull image k8s.gcr.io/kube-controller-manager:v1.13.2: output: Error response from daemon: Get https://k8s.gcr.io/v2/: dial tcp: lookup k8s.gcr.io on [2001:4860:4860::8888]:53: dial udp [2001:4860:4860::8888]:53: connect: network is unreachable
, error: exit status 1
        [ERROR ImagePull]: failed to pull image k8s.gcr.io/kube-scheduler:v1.13.2: output: Error response from daemon: Get https://k8s.gcr.io/v2/: dial tcp: lookup k8s.gcr.io on [2001:4860:4860::8888]:53: dial udp [2001:4860:4860::8888]:53: connect: network is unreachable
, error: exit status 1
        [ERROR ImagePull]: failed to pull image k8s.gcr.io/kube-proxy:v1.13.2: output: Error response from daemon: Get https://k8s.gcr.io/v2/: dial tcp: lookup k8s.gcr.io on [2001:4860:4860::8888]:53: dial udp [2001:4860:4860::8888]:53: connect: network is unreachable
, error: exit status 1
        [ERROR ImagePull]: failed to pull image k8s.gcr.io/pause:3.1: output: Error response from daemon: Get https://k8s.gcr.io/v2/: dial tcp: lookup k8s.gcr.io on [2001:4860:4860::8888]:53: dial udp [2001:4860:4860::8888]:53: connect: network is unreachable
, error: exit status 1
        [ERROR ImagePull]: failed to pull image k8s.gcr.io/etcd:3.2.24: output: Error response from daemon: Get https://k8s.gcr.io/v2/: dial tcp: lookup k8s.gcr.io on [2001:4860:4860::8888]:53: dial udp [2001:4860:4860::8888]:53: connect: network is unreachable
, error: exit status 1
        [ERROR ImagePull]: failed to pull image k8s.gcr.io/coredns:1.2.6: output: Error response from daemon: Get https://k8s.gcr.io/v2/: dial tcp: lookup k8s.gcr.io on [2001:4860:4860::8888]:53: dial udp [2001:4860:4860::8888]:53: connect: network is unreachable
, error: exit status 1
[preflight] If you know what you are doing, you can make a check non-fatal with `--ignore-preflight-errors=...`

: Process exited with status 1
================================================================================
An error has occurred. Would you like to opt in to sending anonymized crash
information to minikube to help prevent future errors?

To disable this prompt, run: 'minikube config set WantReportErrorPrompt false'
================================================================================
Please enter your response [Y/n]: Prompt timed out.
Bummer, perhaps next time!

minikube failed :( exiting with error code 1
```

**Anything else do we need to know**:
I assume I am an edge case and it may simply be the boot2docker image.

**Comment by tstromberg** [comment]
I honestly don't understand enough about this issue to advise further -- but in a nutshell, the VM thinks it's IPV6 enabled, and DNS is responding with an IPV6 IP, so it's trying.

Do you mind outputting the results of `minikube ssh ip a`?

**Comment by BlackMaria** [comment]
Thank for your response. here is the output from `minikube ssh ip a`. So there is ipv4 available :} 

``` bash 
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether f0:5d:df:34:86:f4 brd ff:ff:ff:ff:ff:ff
    inet 192.168.38.74/24 brd 192.168.38.255 scope global dynamic eth0
       valid_lft 3537sec preferred_lft 3537sec
    inet6 fe80::f25d:dfff:fe34:86f4/64 scope link 
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether ec:81:df:93:3c:d5 brd ff:ff:ff:ff:ff:ff
    inet 192.168.39.116/24 brd 192.168.39.255 scope global dynamic eth1
       valid_lft 3537sec preferred_lft 3537sec
    inet6 fe80::ee81:dfff:fe93:3cd5/64 scope link 
       valid_lft forever preferred_lft forever
4: sit0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1000
    link/sit 0.0.0.0 brd 0.0.0.0
5: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default 
    link/ether 02:42:08:1e:0e:51 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever
```

**Comment by afbjorklund** [comment]
It seems like your DNS settings are pointing to some server that it cannot reach ?

```
$ host 2001:4860:4860::8888
8.8.8.8.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.6.8.4.0.6.8.4.1.0.0.2.ip6.arpa domain name pointer google-public-dns-a.google.com.
```

Normally it looks something like this:

``` console
                         _             _            
            _         _ ( )           ( )           
  ___ ___  (_)  ___  (_)| |/')  _   _ | |_      __  
/' _ ` _ `\| |/' _ `\| || , <  ( ) ( )| '_`\  /'__`\
| ( ) ( ) || || ( ) || || |\`\ | (_) || |_) )(  ___/
(_) (_) (_)(_)(_) (_)(_)(_) (_)`\___/'(_,__/'`\____)

$ systemd-resolve --status eth0
Link 2 (eth0)
      Current Scopes: DNS LLMNR/IPv4 LLMNR/IPv6
       LLMNR setting: yes
MulticastDNS setting: no
      DNSSEC setting: no
    DNSSEC supported: no
         DNS Servers: 192.168.122.1
$ 
```

It seems your libvirt config is "different"

**Comment by BlackMaria** [comment]
it would be nice to have a means of forcing ipv4 only.  Noting that minikube creates vm that tries ivp6... unless someone could tell me what was wrong with my installation.

**Comment by terrytibbles** [comment]
Any resolution on this?

**Comment by BlackMaria** [comment]
Nope, but I have moved on to 3ks :)    

However, I suggest that the minikube allow us to insert a start up script. With such I could have hacked this to work,  and another example anyone that requires proxies in the VMs could add those values.
Thanks just the same folks!

**Comment by afbjorklund** [comment]
> However, I suggest that the minikube allow us to insert a start up script. With such I could have hacked this to work, and another example anyone that requires proxies in the VMs could add those values.

There is a startup script in /var/lib/boot2docker

```
# This script will get called when the machine is started up
# Feel free to add any initialization you need to be ran here
if [ -e /var/lib/boot2docker/bootlocal.sh ]; then
    /bin/sh /var/lib/boot2docker/bootlocal.sh
fi
```

The location is inherited from docker-machine
